src/kernel/MomentumOpenAdvDiffElemKernel.C

```diff
@@ -0,0 +1,370 @@
+  get_face_shape_fn_data<BcAlgTraits>([&](double* ptr){meFC->shape_fcn(ptr);}, vf_shape_function_);
+  get_scs_shape_fn_data<BcAlgTraits>([&](double* ptr){meSCS->shape_fcn(ptr);}, v_shape_function_);
+}
+
+template<typename BcAlgTraits>
+MomentumOpenAdvDiffElemKernel<BcAlgTraits>::~MomentumOpenAdvDiffElemKernel()
+{}
+
+template<typename BcAlgTraits>
+void
+MomentumOpenAdvDiffElemKernel<BcAlgTraits>::execute(
+  SharedMemView<DoubleType**> &lhs,
+  SharedMemView<DoubleType *> &rhs,
+  ScratchViews<DoubleType> &faceScratchViews,
+  ScratchViews<DoubleType> &elemScratchViews)
+{
+  DoubleType w_uBip[BcAlgTraits::nDim_];
+  DoubleType w_uScs[BcAlgTraits::nDim_];
+  DoubleType w_uBipExtrap[BcAlgTraits::nDim_];
+  DoubleType w_uspecBip[BcAlgTraits::nDim_];
+  DoubleType w_coordBip[BcAlgTraits::nDim_];
+  DoubleType w_nx[BcAlgTraits::nDim_];
+
+  // face
+  SharedMemView<DoubleType*>& vf_viscosity = faceScratchViews.get_scratch_view_1D(*viscosity_);
+  SharedMemView<DoubleType**>& vf_velocityNp1 = faceScratchViews.get_scratch_view_2D(*velocityNp1_);
+  SharedMemView<DoubleType**>& vf_bcVelocity = faceScratchViews.get_scratch_view_2D(*velocityBc_);
+  SharedMemView<DoubleType***>& vf_Gjui = faceScratchViews.get_scratch_view_3D(*Gjui_);
+  SharedMemView<DoubleType**>& vf_coordinates = faceScratchViews.get_scratch_view_2D(*coordinates_);
+  SharedMemView<DoubleType**>& vf_exposedAreaVec = faceScratchViews.get_scratch_view_2D(*exposedAreaVec_);
+  SharedMemView<DoubleType*>& vf_openMassFlowRate = faceScratchViews.get_scratch_view_1D(*openMassFlowRate_);
+  
+  // element
+  SharedMemView<DoubleType**>& v_coordinates = elemScratchViews.get_scratch_view_2D(*coordinates_);
+  SharedMemView<DoubleType**>& v_velocityNp1 = elemScratchViews.get_scratch_view_2D(*velocityNp1_);
+  SharedMemView<DoubleType**>& v_vrtm = elemScratchViews.get_scratch_view_2D(*velocityRTM_);
+  SharedMemView<DoubleType**>& v_viscosity = elemScratchViews.get_scratch_view_1D(*viscosity_);
+  SharedMemView<DoubleType**>& v_density = elemScratchViews.get_scratch_view_1D(*density_);
+  SharedMemView<DoubleType***>& v_dndx = shiftedGradOp_
+    ? elemScratchViews.get_me_views(CURRENT_COORDINATES).dndx_shifted_fc_scs
+    : elemScratchViews.get_me_views(CURRENT_COORDINATES).dndx_fc_scs;
+
+  // fixme: need master elements
+  MasterElement* meFC = NULL;
+  MasterElement* meSCS = NULL;
+
+  for (int ip=0; ip < BcAlgTraits::numFaceIp_; ++ip) {
+    
+    // fixme: need face_ordinal, hack it in...
+    const int face_ordinal = -1;
+
+    // fixme: need the master element
+    const int opposingNode = meSCS->opposingNodes(face_ordinal,ip);
+    const int nearestNode = ipNodeMap_[ip];
+    const int opposingScsIp = meSCS->opposingFace(face_ordinal,ip);
+    const int localFaceNode = faceIpNodeMap[ip];
+    
+    // fixme: left and right nodes; right is on the face; left is the opposing node
+    stk::mesh::Entity nodeL = stk::mesh::Entity(); // elem_node_rels[opposingNode];
```

Comment:
The nodeL and nodeR can be removed by simply extracting data off of, e..g, v_phi(nearestNode);